cpp/src/search/search.cu

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/search.hpp>
+#include <cudf/legacy/copying.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <thrust/binary_search.h>
+#include <thrust/device_vector.h>
+#include <thrust/logical.h>
+
+namespace cudf {
+namespace experimental {
+
+namespace {
+
+template <typename DataIterator, typename ValuesIterator,
+          typename OutputIterator, typename Comparator>
+void launch_search(DataIterator it_data,
+                   ValuesIterator it_vals,
+                   size_type data_size,
+                   size_type values_size,
+                   OutputIterator it_output,
+                   Comparator comp,
+                   bool find_first,
+                   cudaStream_t stream)
+{
+  if (find_first) {
+    thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+                        it_data, it_data + data_size,
+                        it_vals, it_vals + values_size,
+                        it_output, comp);
+  }
+  else {
+    thrust::upper_bound(rmm::exec_policy(stream)->on(stream),
+                        it_data, it_data + data_size,
+                        it_vals, it_vals + values_size,
+                        it_output, comp);
+  }
+}
+
+} // namespace
+
+namespace detail {
+
+std::unique_ptr<column> search_ordered(table_view const& t,
+                                       table_view const& values,
+                                       bool find_first,
+                                       std::vector<order> const& column_order,
+                                       std::vector<null_order> const& null_precedence,
+                                       cudaStream_t stream,
```

Comment:
Make stream the last parameter and default it to zero, for consistency with every other API (and so you don't have to pass 0).
```suggestion
                                       rmm::mr::device_memory_resource *mr,
                                       cudaStream_t stream = 0)
```